File: jlox/tool/GenerateAst.java
package jlox.tool;

import java.io.IOException;
import java.io.PrintWriter;
import java.util.Arrays;
import java.util.List;

public class GenerateAst {
  public static void main(String[] args) throws IOException {
    if (args.length != 1) {
      System.err.println("Usage: generate_ast <output directory>");
      System.exit(64);
    }

    String outputDir = args[0];

    /**
     * defineAST defines an AST node.
     * 
     * defineAst(output_directory, class_name, Array<Subclasses>)
     * 
     * Each subclass is defined as Subclass_name : <Attr_class : attr_identifier> *
     */

    defineAst(outputDir, "Expr", Arrays.asList(
        "Assign   : Token name, Expr value",
        "Binary   : Expr left, Token operator, Expr right",
        "Grouping : Expr expression",
        "Literal  : Object value",
        "Logical  : Expr left, Token operator, Expr right",
        "This     : Token keyword",
        "Unary    : Token operator, Expr right",
        "Call     : Expr callee, Token paren, List<Expr> arguments",
        "Get      : Expr object, Token name",
        "Set      : Expr object, Token name, Expr value",
        "Super    : Token keyword, Token method",
        "Variable : Token name"));

    defineAst(outputDir, "Stmt", Arrays.asList(
        "Block      : List<Stmt> statements",
        "Class      : Token name, Expr.Variable superclass, List<Stmt.Function> methods",
        "Expression : Expr expression",
        "If         : Expr condition, Stmt thenBranch, Stmt elseBranch",
        "Print      : Expr expression",
        "Return     : Token keyword, Expr value",
        "While      : Expr condition, Stmt body",
        "Var        : Token name, Expr initializer",
        "Function   : Token name, List<Token> params, List<Stmt> body"));
  }

  // Creates a AST object.
  private static void defineAst(
      String outputDir, String baseName, List<String> types)
      throws IOException {
    String path = outputDir + "/" + baseName + ".java";
    PrintWriter writer = new PrintWriter(path, "UTF-8");

    // Generate source for each AST object.
    writer.println("package jlox.lox;");
    writer.println();
    writer.println("import java.util.List;");
    writer.println();
    writer.println("abstract class " + baseName + " {");

    // Generate visitor interface.
    defineVisitor(writer, baseName, types);

    // Generate attributes for each base class.
    for (String type : types) {
      String className = type.split(":")[0].trim();
      String fields = type.split(":")[1].trim();
      defineType(writer, baseName, className, fields);
    }

    // The base accept method
    writer.println();
    writer.println("  abstract <R> R accept(Visitor<R> visitor);");

    writer.println("}");
    writer.close();
  }

  // Define visitor interface
  private static void defineVisitor(
      PrintWriter writer, String baseName, List<String> types) {
    writer.println("  interface Visitor<R> {");

    for (String type : types) {
      String typeName = type.split(":")[0].trim();
      writer.println("    R visit" + typeName + baseName + "(" +
          typeName + " " + baseName.toLowerCase() + ");");
    }

    writer.println("  }");
  }

  // Define a type.
  private static void defineType(
      PrintWriter writer, String baseName,
      String className, String fieldList) {
    writer.println("  static class " + className + " extends " +
        baseName + " {");

    // Constructor.
    writer.println("    " + className + "(" + fieldList + ") {");

    // Store parameters in fields.
    String[] fields = fieldList.split(", ");
    for (String field : fields) {
      String name = field.split(" ")[1];
      writer.println("      this." + name + " = " + name + ";");
    }

    writer.println("    }");

    // Visitor pattern.
    writer.println();
    writer.println("    @Override");
    writer.println("    <R> R accept(Visitor<R> visitor) {");
    writer.println("      return visitor.visit" +
        className + baseName + "(this);");
    writer.println("    }");

    // Fields.
    writer.println();
    for (String field : fields) {
      writer.println("    final " + field + ";");
    }

    writer.println("  }");
  }
}

File: pylox/pylox/generate_ast.py
# pylox/generate_ast.py
import sys


class StringBuilder:
    def __init__(self) -> None:
        self.parts: list[str] = []

    def write(self, x: str) -> None:
        self.parts.append(x)

    def get_string(self) -> str:
        return '\n'.join(self.parts)


def main() -> None:
    if len(sys.argv) != 2:
        print('Usage: python generate_ast.py <output_dir>')
        sys.exit(64)

    output_dir = sys.argv[1]
    write_ast_node(
        output_dir,
        'Expr',
        [
            'Binary - left: Expr, operator: Token, right: Expr',
            'Grouping - expression: Expr',
            'Literal - value: object',
            'Unary - operator: Token, right: Expr',
        ],
    )


def write_ast_node(output_dir: str, base_name: str, defns: list[str]) -> None:
    base_path = f'{output_dir}{base_name.lower()}.py'
    sb = StringBuilder()

    sb.write(f'# {base_path}\n')
    sb.write(""" 
from __future__ import annotations # For forward references

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Protocol, TypeVar
              
from .lox_token import Token
              
T = TypeVar('T', covariant=True)\n
    """)

    sb.write(f'class {base_name}(ABC):')
    sb.write('    @abstractmethod')
    sb.write('    def accept(self, visitor: Visitor[T]) -> T: ...')

    # Write the visitor pattern.
    write_visitor(sb, base_name, defns)

    # The AST classes.
    for defn in defns:
        class_name = defn.split('-')[0].strip()
        fields_str = defn.split('-')[1].strip()
        sb.write('\n')
        write_class(sb, base_name, class_name, fields_str)

    # Write to file
    with open(base_path, 'w') as fp:
        fp.write(sb.get_string() + '\n')


def write_visitor(sb: StringBuilder, base_name: str, defns: list[str]) -> None:
    sb.write('\n')
    sb.write('class Visitor(Protocol[T]):')
    for defn in defns:
        class_name = defn.split('-')[0].strip()
        sb.write(
            f'    def visit_{class_name.lower()}_{base_name.lower()}(self, {base_name.lower()}: {class_name}) -> T: ...'
        )


def write_class(
    sb: StringBuilder, base_name: str, class_name: str, fields_str: str
) -> None:
    sb.write('@dataclass(frozen=True)')
    sb.write(f'class {class_name}({base_name}):')

    for field in fields_str.split(','):
        sb.write(f'    {field.strip()}')

    sb.write('')
    sb.write('    def accept(self, visitor: Visitor[T]) -> T:')
    sb.write(
        f'        return visitor.visit_{class_name.lower()}_{base_name.lower()}(self)'
    )


if __name__ == '__main__':
    main()
